crates/daphne-worker/src/aggregator/router/extractor.rs

```diff
@@ -0,0 +1,794 @@
+            return Err(AxumDapResponse::new_error(
+                fatal_error!(err = "failed to get payload"),
+                state.server_metrics(),
+            ));
+        };
+
+        let taskprov_advertisement =
+            extract_header_as_str(&parts.headers, http_headers::DAP_TASKPROV)
+                .map(|h| TaskprovAdvertisement::parse_taskprov_advertisement(h, &task_id, version))
+                .transpose()
+                .map_err(|e| AxumDapResponse::new_error(e, state.server_metrics()))?;
+
+        let meta = DapRequestMeta {
+            version,
+            task_id,
+            taskprov_advertisement,
+            media_type,
+        };
+
+        let payload = B::decode_from_http_body(payload, &meta)
+            .map_err(|e| AxumDapResponse::new_error(e, state.server_metrics()))?;
+
+        let request = DapRequest {
+            meta,
+            resource_id: resource_parser.into(),
+            payload,
+        };
+
+        Ok(UnauthenticatedDapRequestExtractor(request))
+    }
+}
+
+pub mod dap_sender {
+    pub type DapSender = u8;
+
+    pub const FROM_CLIENT: DapSender = 0;
+    pub const FROM_COLLECTOR: DapSender = 1 << 1;
+    pub const FROM_HELPER: DapSender = 1 << 2;
+    pub const FROM_LEADER: DapSender = 1 << 3;
+
+    pub const fn to_enum(id: DapSender) -> daphne::constants::DapRole {
+        match id {
+            FROM_CLIENT => daphne::constants::DapRole::Client,
+            FROM_COLLECTOR => daphne::constants::DapRole::Collector,
+            FROM_HELPER => daphne::constants::DapRole::Helper,
+            FROM_LEADER => daphne::constants::DapRole::Leader,
+            _ => panic!("invalid dap sender. Please specify a valid dap_sender from the crate::extractor::dap_sender module"),
+        }
+    }
+}
+
+/// An axum extractor capable of parsing a [`DapRequest`].
+///
+/// This extractor asserts that the request is authenticated.
+///
+/// # Type and const parameters
+/// - `SENDER`: The role that is expected to send this request. See [`dap_sender`] for possible values.
+/// - `P` and `R`: See [`DapRequest`] for an explanation these type parameters.
+pub(super) struct DapRequestExtractor<const SENDER: dap_sender::DapSender, B: RequestBody>(
+    pub DapRequest<B>,
+);
+
+#[async_trait]
+impl<const SENDER: dap_sender::DapSender, S, B> FromRequest<S> for DapRequestExtractor<SENDER, B>
+where
+    B: DecodeFromDapHttpBody + Send + Sync,
+    B::ResourceId: resource_parsers::Resource,
+    S: DaphneService + Send + Sync,
+{
+    type Rejection = AxumDapResponse;
+
+    async fn from_request(req: Request, state: &S) -> Result<Self, Self::Rejection> {
+        let bearer_token = extract_header_as_str(req.headers(), http_headers::DAP_AUTH_TOKEN)
+            .map(BearerToken::from);
+
+        let cf_tls_client_auth = mtls_auth_from_request(&req);
+
+        let request = UnauthenticatedDapRequestExtractor::from_request(req, state)
+            .await?
+            .0;
```

Comment:
nit: I find this to be a little neater
```suggestion
        let UnauthenticatedDapRequestExtractor(request) = UnauthenticatedDapRequestExtractor::from_request(req, state)
            .await?;
```